### PR TITLE
Implement np.linalg.{eigvals(), eigh(), eigvalsh()}.

### PR DESCRIPTION
**Pull request by stuartarchibald**
This PR implements:
 * numpy.linalg.eigvals()
 * numpy.linalg.eigh()
 * numpy.linalg.eigvalsh()

and augments the tests and input data for numpy.linalg.eig()
for reuse with the above. Further, it exposes an easy to use
wrapper of the Hermitian eigen-value/vector routines available
from LAPACK.

### CODE REVIEW

File: numba/_lapack.c [hunk 1]
```diff
@@ -983,6 +1046,213 @@ numba_ez_cgeev(char kind, char jobvl, char jobvr,  Py_ssize_t n, void *a,
 }
 
 
+typedef void (*xsyevd_t)(char *jobz, char *uplo, F_INT *n, void *a, F_INT *lda,
```
Comment (pitrou) [inline numba/_lapack.c]: The typedefs are duplicated from above (see line ~574).

File: numba/tests/test_linalg.py [hunk 2]
```diff
@@ -805,25 +817,31 @@ def sample_matrix(self, m, dtype, order):
...
             self.assertEqual(len(expected), len(got))
-            # and that length is 2
-            self.assertEqual(len(got), 2)
+            # and that dimension is correct
+            res_is_tuple = False
+            if isinstance(got, tuple):
+                res_is_tuple = True
+                self.assertEqual(len(got), expected_res_len)
+            else:  # its an array
+                self.assertEqual(got.ndim, expected_res_len)
```
Comment (pitrou) [inline numba/tests/test_linalg.py]: Why not `expected.ndim`?

File: numba/tests/test_linalg.py [hunk 3]
```diff
@@ -805,25 +817,31 @@ def sample_matrix(self, m, dtype, order):
...
             got = cfunc(a)
             # check that the returned tuple is same length
             self.assertEqual(len(expected), len(got))
-            # and that length is 2
-            self.assertEqual(len(got), 2)
+            # and that dimension is correct
+            res_is_tuple = False
+            if isinstance(got, tuple):
+                res_is_tuple = True
+                self.assertEqual(len(got), expected_res_len)
```
Comment (pitrou) [inline numba/tests/test_linalg.py]: Why not `len(expected)`?

Comment (stuartarchibald) [inline numba/tests/test_linalg.py]: It's comparing to a hardcoded known value that is supplied in the arguments to the checker function. There is a comparison to `expected` itself here:
https://github.com/numba/numba/pull/2129/files#diff-3208831c404c0a8b312ff145797b3818R836
I can change/remove this if you wish?

File: numba/tests/test_linalg.py (hunk 2 above)
Comment (stuartarchibald) [inline numba/tests/test_linalg.py]: Same comment as https://github.com/numba/numba/pull/2129#discussion_r80044990

File: numba/_lapack.c (hunk 1 above)
Comment (stuartarchibald) [inline numba/_lapack.c]: Oops, indeed, good spot. I'll remove the dupe.

**Pull request merged**